File: javascript/palindrome-products/palindrome-products.js
//
// This is only a SKELETON file for the 'Palindrome Products' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Palindromes {
  static generate(properties) {
    let found = []

    let isPalindrome = (value) =>
    {
      let str = String(value)
      return (str.length === 1 || str === [... str].reverse().join(''))
    }

    let getFactors = (value, min, max) =>
    {
      let ret = []
      let limit = Math.min(Math.sqrt(value), max)
      for(let i = min; i <= limit; i++)
      {
        if (value % i === 0)
        {          
          let other = value / i
          if (other >= min && other <= max)
            ret.push([i, other])
        }
      }
      return ret
    }

    if (properties.minFactor > properties.maxFactor)
      throw new Error('min must be <= max')

    let min = 0
    let max = 0
    for(let j = properties.minFactor; j <= properties.maxFactor; j++)
    {
      for(let i = j; i <= properties.maxFactor; i++)
      {
        let product = i * j
        if (found.length > 0 && product > min && product < max)
          continue

        if (isPalindrome(product))
        {
          found.push(product)
          if (found.length === 1)
          {
            min = product
            max = product
          }
          else 
          {
            min = Math.min(product, min)
            max = Math.max(product, max)
          }
        }
      }
    }

    if (found.length === 0)
      return { smallest: {value: null, factors:[]}, largest: {value: null, factors: []}}

    let minKey = Math.min(...found)
    let maxKey = Math.max(...found)

    return { 
      smallest: {value: minKey, factors: getFactors(minKey, properties.minFactor, properties.maxFactor)}, 
      largest: {value: maxKey, factors: getFactors(maxKey, properties.minFactor, properties.maxFactor) }}
  }  
}

File: python/palindrome-products/palindrome_products.py
"""
Find the smallest and largest palindrome products for a range of numbers.
"""

def largest(min_factor, max_factor):
    """Given a range of numbers, find the largest palindromes which
       are products of two numbers within that range.

    :param min_factor: int with a default value of 0
    :param max_factor: int
    :return: tuple of (palindrome, iterable).
             Iterable should contain both factors of the palindrome in an arbitrary order.
    """
    return palindrome_product_helper(min_factor, max_factor, max)


def smallest(min_factor, max_factor):
    """Given a range of numbers, find the smallest palindromes which
    are products of two numbers within that range.

    :param min_factor: int with a default value of 0
    :param max_factor: int
    :return: tuple of (palindrome, iterable).
    Iterable should contain both factors of the palindrome in an arbitrary order.
    """
    return palindrome_product_helper(min_factor, max_factor, min)

def palindrome_product_helper(min_factor, max_factor, compare_fn):
    """Given a range of numbers, find the palindromes matching the result of the function fn
    which are products of two numbers within that range.
    :param min_factor: int with a default value of 0
    :param max_factor: int
    :param compare_fn: function to determine palindrome products to keep
    :return: tuple of (palindrome, iterable).
    Iterable should contain both factors of the palindrome in an arbitrary order.
    """
    if min_factor > max_factor:
        raise ValueError('min must be <= max')

    products = []
    target_value = None

    if compare_fn(1, 2) == 2:
        # Bigger first
        start = max_factor
        stop = min_factor - 1
        step = -1
    else:
        # Smaller first
        start = min_factor
        stop = max_factor + 1
        step = 1

    for outer in range(start, stop, step):
        if step > 0:
            inner_start = outer
            inner_stop = max_factor + 1
        else:
            inner_start = max_factor
            inner_stop = outer - 1
        for inner in range(inner_start, inner_stop, step):
            product = inner * outer 
            if target_value is not None and compare_fn(product, target_value) != product:
                continue
            digits = str(inner * outer)
            if digits == digits[::-1]:
                target_value = product
                products.append((inner, outer))

    if len(products) == 0:
        return None, ()
    return target_value, (pair for pair in products if pair[0] * pair[1] == target_value)
